lib/Sema/SemaExpr.cpp

```diff
@@ -13919,14 +13940,26 @@ void Sema::MarkFunctionReferenced(SourceLocation Loc, FunctionDecl *Func,
     } else if (MethodFunName == "__cxxamp_trampoline"||
          MethodFunName == "__cxxamp_trampoline_name") {
       DefineAMPTrampoline(Loc, MethodDecl);
-    } else if (MethodDecl->isOverloadedOperator() &&
-               MethodDecl->getOverloadedOperator() == OO_Equal) {
-      MethodDecl = cast<CXXMethodDecl>(MethodDecl->getFirstDecl());
-      if (MethodDecl->isDefaulted() && !MethodDecl->isDeleted()) {
-        if (MethodDecl->isCopyAssignmentOperator())
-          DefineImplicitCopyAssignment(Loc, MethodDecl);
-        else if (MethodDecl->isMoveAssignmentOperator())
-          DefineImplicitMoveAssignment(Loc, MethodDecl);
+    } else if (MethodDecl->isOverloadedOperator()) {
+      if (is_hip_functor(MethodDecl)) { // TODO: temporary.
+        const auto b = cast<CompoundStmt>(MethodDecl->getBody());
+        auto p = std::find_if(
+            b->body_begin(),
+            b->body_end(),
+            [](Stmt* x) { return isa<CallExpr>(x); });
+        if (p != b->body_end()) {
```

Comment:
That's not quite the assumption, hence the iteration through the body (partially). The invariant is that the ONLY call in the body of the call operator is the call to a __global__ (i.e. [[hc]], there's nothing special about __global__s at the moment) function. Since we opaquely manufacture it in the HIP implementation, nobody can break our invariant externally (without effort, but if they invest the effort they deserve whatever they get). asserting that the callee is actually a __global__ function, while a good idea, entails making __global__ functions more special than they are. I think that we will do some iteration on how to add the "specialness", so perhaps we want to revisit this once we've narrowed that down?